### PR TITLE
Multiple movie support

### PR DESCRIPTION
**Pull request by kmeisthax**
This PR adds support for loading multiple movies into one player, either as separate layers or in specific movie clips. It also supports loading data from variables in `ActionGetURL2`.

Some refactoring was done to `Player` and it's backends along the way; these are independent of the movie loading behavior but were necessary in order to support things we want to do in async code. `wasm_bindgen` does not support spawning futures that borrow linear memory, nor can futures hold garbage-collected pointers. The solution to this is `LoadManager`, a root object which holds all references to GC'd data and contains all async code (so far). It's futures hold an `Arc<Player>` and any other async bits, since those don't rely on the garbage collector and have a lifetime of `'static'`. This motivated removing the type parameters on `Player` as otherwise it's type parameters would leak into `UpdateContext` and thus every other bit of code in the crate.

Loaded SWF movies are encapsulated in a `SwfMovie` type, which holds a parsed header and uncompressed movie data. Each display object holds a reference to the movie it was defined in, either directly or through `SwfSlice`; this reference is stored in the display object's static data. Movies may be loaded in either an arbitrary `MovieClip` or in a level. Levels are implemented by replacing `gc_root.root` with a `BTreeMap` of `MovieClip`s and changing all operations that work on the root to instead visit each layer root in order.

Streaming loads are not yet supported and will be implemented in a future PR. However, now that SWF data is centralized in a single location, streaming should be fairly easy to implement.

# API surfaces to implement

 * [x] `ActionGetUrl` with target `_levelN`
 * [x] `ActionGetUrl2` with target sprite flag
 * [x] `ActionGetUrl2` with load variables flag (loads form data, not movies)
 * ~`getURL`~ - Nevermind. This function does not appear to have movie-loading API surfaces (as opposed to it's opcode equivalents)
 * ~`loadMovie`~ - This doesn't actually exist as a VM callable.
 * ~`loadMovieNum`~ - This doesn't actually exist as a VM callable.
 * ~`loadVariables`~ - This doesn't actually exist as a VM callable.
 * ~`loadVariablesNum`~ - This doesn't actually exist as a VM callable.
 * ~`unloadMovie`~ - This doesn't actually exist as a VM callable.
 * ~`unloadMovieNum`~ - This doesn't actually exist as a VM callable.
 * [x] `MovieClip.loadMovie`
 * [x] `MovieClip.loadVariables`
 * [x] `MovieClip.unloadMovie`
 * [x] `MovieClipLoader`
 * [x] `XML.load` - I don't have a good way to test `send` and `sendAndLoad` so they're out of scope for now.
 * [ ] Move initial source load into core
 * [ ] `MovieClip._url` - will address #325
 * [ ] `MovieClip._lockroot`

# Other tasks

 * [x] impl async for `ruffle_desktop` - right now it just panics. To avoid having to implement Flash sandboxing (separate from the mandatory CORS on `ruffle_web`) we will only support local filesystem loads, same as the test backend.

# Open questions

* [ ] How do we reclaim resources from unloaded movies? Right now, I suspect movies may be cyclically referenced: display objects need to keep their source movie alive via `Arc` but so do library symbols. I tried using `weak_table` hashmaps to ensure libraries are cleaned up but we may need to resort to garbage collecting movies.
* [x] `_root` in AVM is supposed to refer to the current timeline, but cross-movie function calls are not currently accounted for - `_root` is set per action. I don't think this is correct, and it may make more sense to kill `UpdateContext.root` and instead walk the display tree to find it.
  * Root is now dynamically determined. This exposed a small bug with register preloading on cross-movie calls that took three days to diagnose.
* [x] In what order are updates run for layers? I'm currently calling `run_frame` from the lowest layer up.
  * Layers are run in order of creation. If you load two movie clips, even if they load on the same frame, the layer that started it's load first is the first to queue actions, *regardless* of the layer depth order. I don't think any software will care, but it's important to note.

# Known issues

* [x] Cached movie loads sometimes fail to make it to the player. This is noticeable on the homestarrunner.com homepage.

### CODE REVIEW

**Comment by Herschel** [comment]
I had a vague recollection that Flash only supported `_level0` through `9`, but I was wrong, sorry!

Just did some testing and it looks like the level basically gets casted to an `i32`, although the max level seems to be `2130706428`. Not sure why exactly this number and not 2^31... Any higher and it errors/does weird stuff, until `4294967296` where it wraps back around to `_level0`.

**Comment by kmeisthax** [comment]
I'll take the level limit into account... though this means I have to make the list of render roots a `Vec` or `SmallVec` rather than a fixed array.

How does Flash even support ridiculously high `_level` numbers, though? That seems like it'd be allocating gobs of memory to hold an array of pointers to layer roots...

**Comment by Herschel** [comment]
I use a `BTreeMap` for the depth-to-children list in `MovieClip`. Not sure if this is what I'll settle on in the long run, but this should work fine for the `loadMovieNum` levels.

**Comment by kmeisthax** [comment]
I implemented what you suggested and it seems to have fixed the Homestar Runner issue where layers would fail to load.

My next question is: When you load a movie into a clip, how much of the data pre-load remains after? Like, does Flash completely replace the clip object, wipe out user-defined properties, or let the loaded movie totally see whatever got put into it? What happens to child clips?

**Comment by kmeisthax** [comment]
I implemented an async executor that integrates with the Glutin event loop; this allows the desktop version of Ruffle to have async behaviors. (Previously, it would panic as `unimplemented!()`.)

Actually being able to read files is an exercise for another time. Flash Player had a sandbox for loads that I need to implement before we can expose network access to desktop, which is out of scope for this monstrous, foundational PR.

**Comment by swami74983** [comment]
I am waiting for this feature (multiple movie loading based on xml in AVM1). Make it available to us as soon as possible

File: core/src/avm1/globals/movie_clip_loader.rs [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            .as_display_object()
+            .and_then(|dobj| dobj.as_movie_clip())
+        {
+            let ret_obj = ScriptObject::object(context.gc_context, None);
+            ret_obj.define_value(
+                context.gc_context,
+                "bytesLoaded",
+                movieclip
+                    .movie()
+                    .map(|mv| (mv.data().len() + 21).into())
```
Comment (Herschel) [inline core/src/avm1/globals/movie_clip_loader.rs]: This should be the actual length of the compressed SWF file, but this is currently returning the uncompressed length. Also, the + 21 for the header isn't constant -- annoyingly the stage size rect in the header can be variable size (try making a 4000x4000 SWF). I know the size isn't easily available right now -- eventually we'll want to store the file length somewhere in `SwfMovie` or `SwfMovie.header`

**Comment by Herschel** [comment]
Great PR, thanks so much for the patience!

**Pull request merged**